### PR TITLE
Add quick fix all

### PR DESCRIPTION
**Pull request by karlvr**

### CODE REVIEW

**Comment by karlvr** [comment]
I'm going to add another change to this. So it groups all of the checks it finds, assuming that if you've selected a block you want to fix all of the violations in the selection. 

Only problem I'm having is how to describe the fix; I'm planning to use the check name and convert to sentence case.

**Comment by karlvr** [comment]
@jdneo this is maybe an opinionated change, but I think it's a pretty nice experience for the user.

If you've selected one violation, it's the same as it was. If you've selected two you'll see "Fix 2 Checkstyle Final Parameters violations". And if there's more than one fix offered you get a "Fix all 5 Checkstyle violations" as well.

**Comment by karlvr** [comment]
image

**Comment by jdneo** [comment]
This is a very useful feature that I would like to have! Actually we have an issue tracking for this: #179. And I remember that this idea actually came from other lint extension like the TSLint:

image

BTW, seems that the behavior of TSLint changes from when I filed #179. Tt now has only two kinds of quickfix items:
- fix xxxx
- Fix all fixable errors

And `Fix all the `xxx` errors in the file` seems removed at some time in the past. I think we can refer to the same behavior, just provide these two types of quickfix option.

Back to the implementation. One thought about it is that, there are a lot of duplicated codes between `fix one` and `fix all`. Though I haven't implemented by myself, it seems that the server side can always receive a list of errors to be fix. The difference is that, if it's `fix one`, the list will only have one element, and `fix all` will have all the fixable items. I believe the client can handle with such differences.

What do you think?

**Comment by karlvr** [comment]
@jdneo re implementation: yeah, that's a great idea... many of course includes just one.

With linting (eslint in particular) I do always want to fix all issues, so I guess people might just want to fix all with Checkstyle too. I guess I quite liked that it would sort of tell you what sort of issues there are in the selection... but it does seem superfluous. Remove it?

**Comment by jdneo** [comment]
>  I guess people might just want to fix all with Checkstyle too

I guess so too.

So I prefer to making the list as:
- Fix: xxxx
- Fix all auto-fixable Checkstyle violations

People can see what's the violation at the selection by checking the first option, and fix all quickly by clicking the second one. I think that's enough 😀

**Comment by karlvr** [comment]
@jdneo if a user has selected a whole document, if we still have an entry for each individual violation we end up with a _lot_ of quick fixes. Have I understood what you're suggesting correctly? That's why I decided to aggregate all of the similar violations, so my list becomes:

* Fix 4 xxxx
* Fix all auto-fixable Checkstyle violations

**Comment by jdneo** [comment]
@karlvr Got your point! Ok let's go with your proposal:

- Fix xxx (when the number > 1, show the number: Fix 2 xxx)
- Fix all auto-fixable Checkstyle violations

**Comment by karlvr** [comment]
@jdneo ok, check it out now. I've made the standard quick fix command be 1 or many, and I've changed the titles on the quick fix actions.

Here's a visual:
image

File: src/quickFixProvider.ts [hunk 1]
```diff
@@ -1,24 +1,71 @@
...
+                    diagnostics,
+                    command: {
+                        title: 'Fix the Checkstyle violation',
+                        command: CheckstyleExtensionCommands.FIX_CHECKSTYLE_VIOLATIONS,
+                        arguments: [document.uri, [document.offsetAt(diagnostic.range.start)], [diagnostic.code]],
+                    },
+                    kind: CodeActionKind.QuickFix,
+                });
+            } else {
+                codeActions.push({
```
Comment (jdneo) [inline src/quickFixProvider.ts:30]: Can we leverage the implementation in the `else block` to cover the case `diagnostics.length === 1`? Seems feasible to me by introducing a utility method parsing the title according to the diagnostic length?

File: src/quickFixProvider.ts [hunk 2]
```diff
@@ -27,4 +74,45 @@ class QuickFixProvider implements CodeActionProvider {
...
+}
+
 export const quickFixProvider: QuickFixProvider = new QuickFixProvider();
+
+function formatCheckstyleCheck(str: string): string {
+    if (str.endsWith('Check')) {
+        str = str.substring(0, str.length - 'Check'.length);
+    }
+
+    let result: string = '';
```
Comment (jdneo) [inline src/quickFixProvider.ts:102]: Use `startCase` in Lodash? https://lodash.com/docs/4.17.15#startCase

**Comment by jdneo** [comment]
Another issue I found when there are multiple violations at the same place:


https://user-images.githubusercontent.com/6193897/104697344-97265400-574a-11eb-9c65-02954950bcba.mp4

In this case it seems complicated since the AST is changing dynamically. 🤔

**Comment by karlvr** [comment]
@jdneo hmm, yeah, that double problem is a pain. It's of course because the location of the AST node has changed as soon as one visitor has run.

An idea for the solution:

Use two passes. The first pass marks the nodes that are of "interest" to each quick fix by adding them to a property on the AST node using `setProperty`, or some separate `Map`. The second pass applies changes, and if it splits a node into two, it can copy the "of interest" list for the node so that the other quick fixes can still follow. If a fix copies a subtree, then if we use AST properties, we should also copy the "of interest" property so it should "just work".

That feels like it might be robust. The first pass is read-only and lets us escape from the position-based logic, which is brittle.

**Comment by jdneo** [comment]
Forgive me that I'm not an expert on the AST related APIs. 😄  Not quite get your idea.

Could you help me understand:

> it can copy the "of interest" list for the node 

What's the meaning for `of interest` here?

> If a fix copies a subtree

Example about copy a subtree?

**Comment by karlvr** [comment]
@jdneo no problem, neither am I. The `ASTNode` has a `setProperty` method, that appears to be designed for people like us to add metadata to the tree for our own purposes. By "of interest" I meant the nodes that each quick fix wants to modify; we currently identify those nodes by visiting every node of a given type in the document, and choosing the ones that overlap our Checkstyle marker's position. Copies a subtree occurs when the `copy` method is called in `BaseQuickFix`. For example the `MultipleVariableDeclarationsQuickFix` copies the `VariableDeclarationFragment` and some other nodes (all of which might actually be leaves in the tree, rather than subtrees).

With all that said, I've tried an alternative approach. I've set the `startPosition` and `length` of each created or copied node to match the original. That way when one quick fix runs after another, it can still match newly created nodes. This appears to work... I've added a check that we add this to all of the nodes we're interested in (as I couldn't find a way to add it to all copied nodes easily), and we need to make sure that the quick fix visitor methods are OK visiting any new nodes they create (which happens) so we don't end up in an infinite loop.

**Comment by jdneo** [comment]
The new implementation looks promising. Another feedback is that, in TSLint, when I click `fix all`, it will fix all the fixable violations in the whole document. Here it seems that we are fixing the violations at the cursor position?

**Comment by karlvr** [comment]
@jdneo yeah, that's right, it just fixes what's in the selection! I'll take a look and see if it's easy to expand to the whole document... perhaps if there's no selection (just a cursor) then it offers to fix whatever is under the cursor, and the whole document. But if there is a selection it offers to fix everything in what's selected like it does now?

**Comment by jdneo** [comment]
> if there's no selection (just a cursor) then it offers to fix whatever is under the cursor, and the whole document. But if there is a selection it offers to fix everything in what's selected like it does now?

Sounds good to me.

**Comment by karlvr** [comment]
@jdneo The `CodeActionProvider` that we're currently using seems to only provide access to the `Diagnostic`s in the selected range. I don't see a way to get access to all of the diagnostics for the whole document. The way that our quick fixes are implemented is to use the diagnostics that we've already provided to VS Code to drive the fixes we apply.

I guess we would need to run the checker ourselves against the document and then use those results internally. I'll check to see how that would work...

**Comment by karlvr** [comment]
That turned out to be easy. We can fetch all of the known `Diagnostic`s from our diagnostics collector. Stand by.

**Comment by karlvr** [comment]
@jdneo OK, I've pushed a solution. I've kept the "fix all in selection" if the user has made a selection (rather than just a cursor), but always show the fix all in document. See what you think.

**Review by jdneo — APPROVED** [review]
Nice job! Thank you @karlvr!

**Comment by jdneo** [comment]
Just FYI, I plan to make a new release(1.4.0), since we have some good progress since last version. Kudos to @karlvr !

**Pull request merged**